lib/Dialect/FIRRTL/Transforms/GrandCentral.cpp

```diff
@@ -1599,6 +1644,52 @@ void GrandCentralPass::runOnOperation() {
     LLVM_DEBUG({ llvm::dbgs() << "Generated YAML:" << yamlString << "\n"; });
   }
 
+  // Garbage collect dead NLAs.
+  for (auto &op :
+       llvm::make_early_inc_range(circuitOp.getBody()->getOperations())) {
+
+    // Remove NLA operations.
+    if (auto nla = dyn_cast<NonLocalAnchor>(op)) {
+      if (deadNLAs.count(nla.sym_nameAttr()))
+        nla.erase();
+      continue;
+    }
+
+    auto fmodule = dyn_cast<FModuleOp>(op);
+    if (!fmodule)
+      continue;
+
+    // Visit module bodies to remove any dead NLAs.
+    for (auto &op : *fmodule.getBody()) {
+      AnnotationSet annotations(&op);
+      if (annotations.empty())
+        continue;
+
+      SmallVector<Attribute> newAnnotations;
+      auto isDead = [&](Annotation anno) -> bool {
+        auto sym = anno.getMember<FlatSymbolRefAttr>("circt.nonlocal");
+        if (!sym)
+          return false;
+        return deadNLAs.count(sym.getAttr());
+      };
+
+      annotations.removeAnnotations(isDead);
+      annotations.addAnnotations(newAnnotations);
+      annotations.applyToOperation(&op);
+
+      SmallVector<Attribute> newPortAnnotations;
+      for (auto port : fmodule.getPorts()) {
+        newAnnotations.clear();
+        port.annotations.removeAnnotations(isDead);
+        port.annotations.addAnnotations(newAnnotations);
+        newPortAnnotations.push_back(
+            ArrayAttr::get(op.getContext(), port.annotations.getArray()));
+      }
+      fmodule->setAttr("portAnnotations",
+                       ArrayAttr::get(op.getContext(), newPortAnnotations));
+    }
```

Comment:
Maybe use `bool AnnotationSet::removePortAnnotations(`  instead

Reply:
Looking at this again, this whole thing is totally unnecessary. This code is supposed to be removing breadcrumbs, but it's visiting every op and visiting module ports (one for every op in the module!). This can be entirely removed. I've change this outer loop to just visiting instance ops (and skip ports).